File: Sean的leetcode/Leetcode/Remove Invalid Parentheses.cpp
/* DFS
 * Limit max removal rmL and rmR for backtracking boundary. Otherwise it will exhaust all possible valid substrings, not shortest ones.
 * Scan from left to right, avoiding invalid strs (on the fly) by checking num of open parens.
 * If it's '(', either use it, or remove it.
 * If it's '(', either use it, or remove it.
 * Otherwise just append it.
*/
class Solution {
public:
    vector<string> removeInvalidParentheses(string s) {
        int l_rm = 0, r_rm = 0;
        unordered_set<string> res;
        int size = s.size();
        for (int i = 0; i < size; i++) {
            if (s[i] == '(') {
                l_rm++;
            }
            else if (s[i] == ')') {
                if (l_rm != 0) {
                    l_rm--;
                }
                else {
                    r_rm++;
                }
            }
        }
        dfs(s, 0, l_rm, r_rm, 0, "", res);
        return vector<string> (res.begin(), res.end());
    }
    void dfs(string s, int idx, int l_rm, int r_rm, int stack, string path, unordered_set<string>& res) {
       
        int size = s.size();
        if (idx == size) {
            if (l_rm == 0 && r_rm == 0 && stack == 0) {
                res.insert(path);
            }
            return;
        }
       
        if (s[idx] == '(') {
            if (l_rm){
                dfs(s, idx + 1, l_rm - 1, r_rm, stack, path, res);
            }
            dfs(s, idx + 1, l_rm, r_rm, stack + 1, path + s[idx], res);
        }
        else if (s[idx] == ')') {
            if (r_rm){
                dfs(s, idx + 1, l_rm, r_rm - 1, stack, path, res);
            }
            if (stack) {
                dfs(s, idx + 1, l_rm, r_rm, stack - 1, path + s[idx], res);
            }
        }
        else{
            dfs(s, idx + 1, l_rm, r_rm, stack, path + s[idx], res); 
        }
       
    }
};

File: Sean的leetcode/FB/Remove Invalid Parentheses.py
###
# DFS
# Time Complexity: O(2^n)
# Space Complexity: O(n)
###
class Solution(object):
    def removeInvalidParentheses(self, s):
        """
        :type s: str
        :rtype: List[str]
        """
        if not s:
            return [""]
        l_rm = r_rm = 0
        for c in s:
            if c == '(':
                l_rm += 1
            elif c == ')':
                if l_rm:
                    l_rm -= 1
                else:
                    r_rm += 1
        
        re = set()
        self.dfs(s, 0, 0, l_rm, r_rm, "", re)
        return list(re)
    
    def dfs(self, s, idx, pair, l_rm, r_rm, one_sol, re):
        if idx == len(s):
            if not pair and not l_rm and not r_rm:
                re.add(one_sol)
            return
        c = s[idx]
        if c == '(':
            self.dfs(s, idx + 1, pair + 1, l_rm, r_rm, one_sol + "(", re)
            if l_rm:
                self.dfs(s, idx + 1, pair, l_rm - 1, r_rm, one_sol, re)
        elif c == ')':
            if pair:
                self.dfs(s, idx + 1, pair - 1, l_rm, r_rm, one_sol + ")", re)
            if r_rm:
                self.dfs(s, idx + 1, pair, l_rm, r_rm - 1, one_sol, re)
        else:
            self.dfs(s, idx + 1, pair, l_rm, r_rm, one_sol + c, re)
            
            